tracer/src/Datadog.Trace/AppSec/Security.cs

```diff
@@ -281,60 +278,112 @@ private void SetRemoteConfigCapabilites()
+            }
         }
 
         private void AsmDataProductConfigChanged(object sender, ProductConfigChangedEventArgs e)
         {
-            if (!_enabled)
-            {
-                return;
-            }
+            if (!_enabled) { return; }
 
-            _asmDataConfigs ??= new Dictionary<string, Payload>();
-            var asmDataConfigs = e.GetDeserializedConfigurations<Payload>();
-            foreach (var asmDataConfig in asmDataConfigs)
+            lock (_settings)
             {
-                _asmDataConfigs[asmDataConfig.Name] = asmDataConfig.TypedFile;
-                e.Acknowledge(asmDataConfig.Name);
-            }
+                _asmDataConfigs ??= new Dictionary<string, RcmModels.AsmData.Payload>();
+                var asmDataConfigs = e.GetDeserializedConfigurations<RcmModels.AsmData.Payload>();
+                foreach (var asmDataConfig in asmDataConfigs)
+                {
+                    _asmDataConfigs[asmDataConfig.Name] = asmDataConfig.TypedFile;
+                    e.Acknowledge(asmDataConfig.Name);
+                }
 
-            var updated = UpdateRulesData();
-            foreach (var asmDataConfig in asmDataConfigs)
-            {
-                if (!updated)
+                var updated = UpdateRulesData();
+                foreach (var asmDataConfig in asmDataConfigs)
                 {
-                    e.Error(asmDataConfig.Name, "Waf could not update the rules");
+                    if (!updated)
+                    {
+                        e.Error(asmDataConfig.Name, "Waf could not update the rules");
+                    }
+                    else
+                    {
+                        e.Acknowledge(asmDataConfig.Name);
+                    }
                 }
-                else
+            }
+        }
+
+        private void AsmProductConfigChanged(object sender, ProductConfigChangedEventArgs e)
+        {
+            if (!_enabled) { return; }
+
+            lock (_settings)
+            {
+                _ruleStatus ??= new Dictionary<string, bool>(StringComparer.InvariantCultureIgnoreCase);
+                var asmConfigs = e.GetDeserializedConfigurations<RcmModels.Asm.Payload>();
+                int ruleCount = 0;
+                foreach (var asmConfig in asmConfigs)
                 {
-                    e.Acknowledge(asmDataConfig.Name);
+                    try
+                    {
+                        var rulesStatus = asmConfig.TypedFile.RuleStatus;
+                        foreach (var data in rulesStatus)
+                        {
+                            if (data.Id == null || data.Enabled == null)
+                            {
+                                continue;
+                            }
+
+                            _ruleStatus[data.Id] = data.Enabled.Value;
+                            ruleCount++;
+                        }
+
+                        if (ruleCount > 0)
```

Comment:
not sure this is the correct behaviour, it would seem strange to receive a null rule id or enabled flag, so may we should warn if any or the data has to be ignored.

Reply:
I agree